### PR TITLE
Bank account verification integration

### PR DESCRIPTION
**Pull request by PaulThor**

### CODE REVIEW

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/connectors/BavfConnector.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+class BavfConnector @Inject()(httpClient: HttpClient)(implicit val appConfig: AppConfig) {
+
+  def init(continueUrl: String,
+           messages: Option[InitRequestMessages] = None,
+           customisationsUrl: Option[String] = None)(
+            implicit ec: ExecutionContext,
+            hc: HeaderCarrier
+          ): Future[Option[InitResponse]] = {
+    import HttpReads.Implicits.readRaw
+    import InitRequest.writes
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/connectors/BavfConnector.scala:24]: personally I'd put these imports at the top of the file with the others.

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala [hunk 2]
```diff
@@ -29,14 +29,22 @@ class MonthsController @Inject()(
...
-        }
-      }).getOrElse(Future.successful(Ok("Session id not set")))
+  val get: Action[AnyContent] = auth.authoriseWithJourneySession { implicit request => vrn => journeySession =>
+
+    journeySession.outStandingAmount match {
+      case Some(outStandingAmount) => Future.successful(Ok(howManyMonthsPage(getMonths(outStandingAmount))))
+      case _ => Future.successful(Redirect(routes.DeferredVatBillController.get()))
+    }
+  }
+
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala:39]: You could do - 
```
    Future.successful(
      journeySession.outStandingAmount match {
        case Some(outStandingAmount) => Ok(howManyMonthsPage(getMonths(outStandingAmount)))
        case _ => Redirect(routes.DeferredVatBillController.get())
      }
    )
```

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala [hunk 3]
```diff
@@ -29,14 +29,22 @@ class MonthsController @Inject()(
...
+  val post: Action[AnyContent] = auth.authoriseWithJourneySession { implicit request => vrn => journeySession =>
+
+      val months = request.body.asFormUrlEncoded.map(a => a.mapValues(_.last)).flatMap(b => b.get("how-many-months"))
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala:42]:
```
request.body.asFormUrlEncoded.flatMap(_.mapValues(_.last).get("how-many-months"))
```

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala (hunk 2 above)
Comment (hmrclt) [reply app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/MonthsController.scala:39]: You may also want to consider writing the auth logic using an ActionBuilder. That way you would have a synchronous and asynchronous method to call on.

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/WhenToPayController.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    (implicit val appConfig: AppConfig, val serviceConfig: ServicesConfig)
+    extends FrontendController(mcc) with I18nSupport {
+
+  val get: Action[AnyContent] = auth.authoriseWithJourneySession { implicit request => vrn => journeySession =>
+
+    journeySession.outStandingAmount match {
+      case Some(_) => Future.successful(Ok(whenToPagePage()))
+      case _ => Future.successful(Redirect(routes.DeferredVatBillController.get()))
+    }
+  }
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/WhenToPayController.scala:41]: As above, you can move the Future.successful to the outside, but I suspect the ideal approach would be to use an actionbuilder here.

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/Enumerable.scala [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+package uk.gov.hmrc.vatdeferralnewpaymentschemefrontend.model.Bavf
+
+import play.api.libs.json._
+
+trait Enumerable[A] {
+
+  def withName(str: String): Option[A]
+}
+
+object Enumerable {
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/Enumerable.scala:15]: You could consider the enumeratum library here.

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/InitRequestAddress.scala [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ */
+
+package uk.gov.hmrc.vatdeferralnewpaymentschemefrontend.model.Bavf
+
+case class InitRequestAddress(lines: List[String], town: Option[String], postcode: Option[String])
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/InitRequestAddress.scala:8]: Is this not the same thing as CompleteResponseAddress?

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/views/WhenToPayPage.scala.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+@()(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
+
+@layout(pageTitle = Some("vat-deferral-new-payment-scheme-frontend")) {
+
+    <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds">
+            <h1 class="govuk-heading-l">@messages("when.to.pay.heading")</h1>
+
+
+            <form action="" method="POST">
```
Comment (hmrclt) [inline app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/views/WhenToPayPage.scala.html:21]: is the action intentionally blank here?

**Review by hmrclt — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/WhenToPayController.scala (hunk 4 above)
Comment (PaulThor) [reply app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/controllers/WhenToPayController.scala:41]: Might need to look into action builders, but will leave it for now

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/Enumerable.scala (hunk 5 above)
Comment (PaulThor) [reply app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/Enumerable.scala:15]: Ok - need to have a look into that, I took this code from the bank-account-verification-frontend-example project so I will look at improving it at a later date

File: app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/InitRequestAddress.scala (hunk 6 above)
Comment (PaulThor) [reply app/uk/gov/hmrc/vatdeferralnewpaymentschemefrontend/model/Bavf/InitRequestAddress.scala:8]: I took this code from bank-account-verification-example-frontend so I'll look into at a later date

**Pull request merged**